# origin: torvalds (BitKeeper)
# cset: 1.971.58.39 (2.6) key=3e868ffdWSVmjpKhisQ9Y-t4YQuXkg (Modified)
# inclusion: upstream
# descrition: Update direct-rendering to current DRI CVS tree.
# revision date: Fri, 07 Jan 2005 16:14:09 +0900
#
# Note this patch has been modified. All components eccept for the addition
# of LOCK_TEST_WITH_RETURN drivers/char/drm/drmP.h have been removed.
# This portion has been updated to correlate locks using pid instead
# of filp as the latter does not exist in 2.4.27. This idea was taken from
# http://mirror.clarkson.edu/pub/distributions/gentoo-portage/sys-kernel/gentoo-sources/files/gentoo-sources-2.4.20-CAN-2004-1056.patch
# -- Horms 7th January 2004
#
# S rset: ChangeSet|1.971.58.38..1.971.58.39
# D rset: drivers/char/drm/drm_agpsupport.h|1.15..1.16
# D rset: drivers/char/drm/i830_dma.c|1.12..1.13
# D rset: drivers/char/drm/drm_proc.h|1.8..1.9
# D rset: drivers/char/drm/r128_cce.c|1.9..1.10
# D rset: drivers/char/drm/r128_state.c|1.10..1.11
# D rset: drivers/char/drm/radeon_cp.c|1.15..1.16
# D rset: drivers/char/drm/drm_bufs.h|1.9..1.10
# D rset: drivers/char/drm/drm_ioctl.h|1.8..1.9
# D rset: drivers/char/drm/i830.h|1.5..1.6
# D rset: drivers/char/drm/i810.h|1.5..1.6
# D rset: drivers/char/drm/i810_dma.c|1.20..1.21
# D rset: drivers/char/drm/i830_drm.h|1.5..1.6
# D rset: drivers/char/drm/mga_state.c|1.12..1.13
# D rset: drivers/char/drm/gamma_drv.h|1.5..1.6
# D rset: drivers/char/drm/radeon_drv.h|1.17..1.18
# D rset: drivers/char/drm/drm_fops.h|1.8..1.9
# D rset: drivers/char/drm/drm_dma.h|1.10..1.11
# D rset: drivers/char/drm/radeon_drm.h|1.11..1.12
# D rset: drivers/char/drm/drm_lock.h|1.6..1.7
# D rset: drivers/char/drm/sis_mm.c|1.4..1.5
# D rset: drivers/char/drm/i830_drv.h|1.6..1.7
# I rset: drivers/char/drm/drmP.h|1.17..1.18
# D rset: drivers/char/drm/drm_drv.h|1.14..1.15
# D rset: BitKeeper/deleted/.del-drm_lists.h~8e5aa1ea5d68c6c1|1.6..1.7
# D rset: drivers/char/drm/radeon_state.c|1.18..1.19
# D rset: drivers/char/drm/Kconfig|1.2..1.3
# D rset: drivers/char/drm/radeon_mem.c|1.5..1.6
# D rset: drivers/char/drm/mga_dma.c|1.10..1.11
# D rset: drivers/char/drm/mga_drv.h|1.11..1.12
# D rset: drivers/char/drm/radeon.h|1.9..1.10
# D rset: drivers/char/drm/drm_os_linux.h|1.7..1.8
# D rset: drivers/char/drm/gamma_dma.c|1.8..1.9
# D rset: drivers/char/drm/radeon_irq.c|1.7..1.8
# D rset: drivers/char/drm/i810_drv.h|1.7..1.8
# D rset: drivers/char/drm/Makefile|1.15..1.16
# D rset: drivers/char/drm/sis.h|1.4..1.5
# D rset: drivers/char/drm/r128_drv.h|1.12..1.13
# D rset: drivers/char/drm/i830_irq.c|1.0..1.1
#
# Key:
# S: Skipped  ChangeSet file only
# O: Original Followed by Updated
# U: Updated  Included with updated range of versions
# I: Included Included verbatim
# E: Excluded Excluded on request from user
# D: Deleted  Manually deleted by subsequent user edit
#
# This is a BitKeeper generated diff -Nru style patch.
#
# ChangeSet
#   2003/03/29 22:34:37-08:00 torvalds@home.transmeta.com 
#   Update direct-rendering to current DRI CVS tree.
#   
#   This adds support for i830 interrupt handling, and new improved
#   lock context keying. See per-file comments for more detail, as this
#   commit sadly mixes up a few different things (that's what you get
#   for not tracking the changes at a fine enough granularity).
# 
# drivers/char/drm/drmP.h
#   2003/03/29 22:34:33-08:00 torvalds@home.transmeta.com +20 -5
#   Make the DRI locking use "struct file" as the fundamental identity of
#   a direct-rendering context. This fixes various leaks, and makes sure
#   that we don't get confused by the file descriptor being released.
#
--- 1.17/drivers/char/drm/drmP.h	2002-12-19 02:56:28 +09:00
+++ 1.18/drivers/char/drm/drmP.h	2003-03-30 15:34:33 +09:00
@@ -268,6 +269,17 @@
 	(_map) = (_dev)->context_sareas[_ctx];		\
 } while(0)
 
+#define LOCK_TEST_WITH_RETURN( dev, filp )				\
+do {									\
+	if ( !_DRM_LOCK_IS_HELD( dev->lock.hw_lock->lock ) ||		\
+	     dev->lock.pid != current->pid ) {				\
+		DRM_ERROR( "%s called without lock held\n",		\
+			   __FUNCTION__ );				\
+		return -EINVAL;						\
+	}								\
+} while (0)
+
+
 typedef int drm_ioctl_t( struct inode *inode, struct file *filp,
 			 unsigned int cmd, unsigned long arg );
 
